Add block spacing and layout to Post Template

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #44899 and  #44557.

* Converts the custom "grid" layout in the Post Template block to use the grid layout type from #49018.
* The grid is configured to be static with a user-defined number of columns and a identical responsive behaviour to the current custom styes.
* The controls to toggle between List and Grid layouts are moved from Query into Post Template toolbar.
* Adds block spacing support to Post Template.

Note: the deprecation had to be done at Query level, because that's where the legacy layout type is stored. This doesn't really work on the server side though, so I opted to leave legacy classnames and styles for unchanged blocks on the front end.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Go to the Home template in a block theme and check that Query and Post Template display as expected on this branch: should look exactly the same as on trunk before any changes are made. 
2. Set the layout to Grid in Post Template toolbar, and change the number of columns in the sidebar. This should work the same as on trunk.
3. Change block spacing from the Post Template block sidebar; check that it works for both the list and the grid layout options.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

tellthemachines committed:
Add static column number option

tellthemachines committed:
Address feedback and fix Group placeholder

tellthemachines committed:
Try using grid layout in Post Template

tellthemachines committed:
Revert non-responsive option

tellthemachines committed:
Update Post Template to use auto-fill columns

tellthemachines commented:
Updated this draft to use column width instead of number of columns, to align better with #49018. If this approach is good, I can start working out deprecation details soon 😅

richtabor commented:
We should use the same control that's on the new grid layout (and it should probably go in the "Layout" panel: 

<img width="277" alt="CleanShot 2023-03-28 at 15 10 11" src="https://user-images.githubusercontent.com/1813435/228342688-a330d12b-3daa-443e-b38d-dc83bdb3e6b5.png">

And it seems funky having the Layout Inner blocks controls with the grid controls—the grid block does not have these: 

<img width="280" alt="CleanShot 2023-03-28 at 15 12 45" src="https://user-images.githubusercontent.com/1813435/228343330-00207f77-7e99-4142-9126-6c34e5908f8b.png">

Is there a cheap way to inherit that same Layout panel from the Grid block? It'll continue to get enhancements moving forward as well.

richtabor commented:
Perhaps the control should fallback to a default column width value, if switched from list to grid. Would prevent cases where you can't tell that it gridded, if the column width was large:

https://user-images.githubusercontent.com/1813435/228346372-0f89fa62-a3c3-4e21-81af-e3c7aa41b096.mp4

tellthemachines commented:
> We should use the same control that's on the new grid layout (and it should probably go in the "Layout" panel:
>
> And it seems funky having the Layout Inner blocks controls with the grid controls—the grid block does not have these:

The problem here is that the Query block itself has a flow/constrained layout so it can set the content width of its children, and it's using custom logic to toggle between the list and the grid display, but those list/grid styles are actually applied to the child Post Template block. To complicate things further, Post Template itself has a locked flow layout type, probably in order to leverage the default theme block spacing value when it's set to list display.

Untangling all this is going to be complicated! It helps to be clear about the ideal state we want to move towards, so with that in mind here are some questions:

* Do we want to preserve the ability to set content width in the Query block? If so, we'll need some controls there for that purpose.
* Would we consider moving the list/grid display controls into the Post Template block where they belong? This has the disadvantage of forcing users to switch between blocks in order to configure the display, so I'm guessing it's a no 😅 but at the same time it's a bit weird that layout controls are all in Query and style controls (colors, typography) are all in Post Template.
* As per #44899 we want to make block spacing configurable in Post Template, but given that Post Template's layout related controls currently live in Query, should the block spacing control be in Query too? This will probably require some re-wiring because block supports are meant to apply to whichever block they are set on.

Ultimately the important thing is deciding whereabouts in this interdependent nested block structure we want the controls to display; I think it would make sense to have them all together. 

It would greatly simplify the interface if Post Template could be invisible and non-configurable, but at this point I'm not sure we can change that without breaking back compat 🤔

tellthemachines commented on packages/block-library/src/post-template/edit.js:
```diff
@@ -95,11 +96,23 @@ export default function PostTemplateEdit( {
+	const updatedLayoutType =
+		layoutType === 'list' || layoutType === 'default' ? 'default' : 'grid';
+
+	useEffect( () => {
```
I don't love this but don't think there's a way around it unless we move the list/grid controls into Post Template instead of keeping them in Query.

andrewserong replied:
It might be worth looking into exploring moving them down 🤔

tellthemachines committed:
Add block spacing to post template

tellthemachines commented:
Looking at ways to leverage this work for #44899; I've now added block spacing support to the Post Template block. Controls can be found in the Post Template sidebar, under "Dimensions".

tellthemachines committed:
Show gap control by default

andrewserong commented:
Very interesting questions and problems to solve! Some first impressions from testing it out:

> Would we consider moving the list/grid display controls into the Post Template block where they belong?

Personally, I'd expect the controls to exist at the Post Template block level, though it might be challenging for folks to discover. The Query block has a lot of complexity to the settings with options for post type, ordering, filters, etc, so I quite like the idea of the styling being deferred to the Post Template level, so that there's a semantic difference there (Query is where you manage the query, Post Template is where you manage the appearance). But that could just be me!

> Is it acceptable to change the "grid" layout setting in Query loop to "minimum column width" instead of "number of colums"? (We could potentially keep "number of columns" by adding further logic to the grid layout type to handle that case)

I feel like in an ideal case we might have both options. At the moment, switching to minimum column width could be perceived as a regression depending on the pattern or layout in use, since it's a widely used block and there are plugins that depend on it (e.g. WooCommerce). For example, here's how the default grid / 3 column pattern appears on trunk versus this PR when using a wide alignment:

![image](https://user-images.githubusercontent.com/14988353/233517714-7aabfda1-9732-494c-a180-46b2d471c9d7.png)

Another thought: for backwards compatibility, are there any existing aspects of using `flex` that will be lost with switching to `grid`? We currently don't have alignment controls exposed, but I wonder if there's a case to be made for also having flex layout exposed in the future so that folks could use either `flex` or `grid` layouts 🤔

tellthemachines commented:
Thanks for the feedback @andrewserong !

> I feel like in an ideal case we might have both options.

Sure, but for now I'll settle whatever gets us to a mergeable solution without too much back and forth around how to materialise those options in the UI 😅 

> here's how the default grid / 3 column pattern appears on trunk versus this PR when using a wide alignment

Hmm, the default column width could be changed to match the 3 col default better, but yeah it will always behave differently.

> Another thought: for backwards compatibility, are there any existing aspects of using flex that will be lost with switching to grid? 

Not with the current implementation of flex in Post Template. In the future, for most alignment options we can rely on grid for the same features as flex; I think flex could only really be useful in super edge case situations such as having only 2 posts in the query and wanting them to be spaced-between? But that seems like a pretty unlikely use case.

andrewserong commented:
> Sure, but for now I'll settle whatever gets us to a mergeable solution without too much back and forth around how to materialise those options in the UI 😅

Good point, it'd be good to try to keep the change as simple as possible if we can to avoid scope creep. In that case, would it be worth exploring retaining the number of columns behaviour, even if it is a bit different to the minimum column width controls in the Grid block? I'm mostly wondering if that would help with the backwards compatibility concern for blocks that might already have a custom column count set.

tellthemachines committed:
Add unstable columns grid property

tellthemachines commented:
I added an unstable "columns" property to layout and some extra logic to make it work. Given the experimental nature of grid (and layout in general 😅 ) this shouldn't work against future experimentation.

That solves the back compat issue, all except for the responsive behaviour that isn't working yet. I'm undecided on whether to add some custom logic to Post Template or to add it as part of layout support; if we ever want to explicitly support defining a fixed number of columns we'll probably want it to have some responsivity.

 I guess the main decision left to make is whether to leave the layout controls where they are or bring them down into Post Template.

Deprecations etc. to be dealt with once we're clear on how much to change here 😄

andrewserong commented:
> Deprecations etc. to be dealt with once we're clear on how much to change here 😄

They can be very fiddly, so good idea to hold off until it's clear which direction to go in!

> I added an unstable "columns" property to layout and some extra logic to make it work.

This looks like a good approach to me! One quick question: since these values will be stored in post content, would it help in the long-term to give it something closer to a final name, possibly `columnCount` — that way if and when we want the attribute to no longer be unstable, we don't have to have code paths to convert the name? Or do we want to reserve that namespace so that `__unstableColumns` is treated explicitly as something to deprecate in code?

tellthemachines commented:
> One quick question: since these values will be stored in post content, would it help in the long-term to give it something closer to a final name, possibly `columnCount` — that way if and when we want the attribute to no longer be unstable, we don't have to have code paths to convert the name? Or do we want to reserve that namespace so that `__unstableColumns` is treated explicitly as something to deprecate in code?

Good question. My main aim for now is to signal that this property could change and shouldn't be relied on. But yeah it's worth making the not-`unstable` part of the name whatever we want it to be if this property becomes stable, as that's how it's usually done elsewhere in the codebase. `columnCount` is a good option!

tellthemachines commented:
> I'm undecided on whether to add some custom logic to Post Template or to add it as part of layout support

Adding responsive rules as part of layout will likely require modifying `wp_style_engine_get_stylesheet_from_css_rules` et al to deal with media queries in some way 🤔

tellthemachines committed:
Rename layout column attribute.

tellthemachines committed:
Add a media query to reproduce current responsiveness

tellthemachines committed:
Move layout controls to Post Template block

tellthemachines commented on packages/block-library/src/post-template/style.scss:
```diff
@@ -30,3 +30,9 @@
 }
+
+@media ( max-width: $break-small ) {
+	.wp-block-post-template-is-layout-grid.wp-block-post-template-is-layout-grid.wp-block-post-template-is-layout-grid.wp-block-post-template-is-layout-grid {
```
Going with a block-specific media query for now to copy the current responsive behaviour. This selector is _far from_ ideal though. I wonder if, now that layout rule specificity has been reduced, it would also be possible to reduce specificity for the `wp-container-x` selector? Ideally though, we'd update the style engine to process media queries so they can be added as part of the layout support (also in the future we should look into moving to container queries).

ramonjd replied:
> Ideally though, we'd update the style engine to process media queries so they can be added as part of the layout support

This is an interesting thought. I think it shouldn't be too hard to support media/container queries in the future.

I've added an item to the [style engine project board](https://github.com/orgs/WordPress/projects/19).

tellthemachines commented:
I've moved the layout controls to the Post Template block as per comments in #44899; also updated the columnCount variable name and cleaned up a bit. I think this is starting to take shape, but would appreciate further feedback before going down the deprecation rabbit hole 😅

andrewserong reviewed (commented):
That's feeling much nicer to me having the controls at the Post Template level 👍

One challenge with the column count approach appears to be that the container can overflow if the value is set too high. Is there a way we can introduce some kind of `minmax` rule at all to determine a cap on how many columns can be displayed at once depending on the size of the container? I wasn't sure how flexible we can be with `repeat` rules, but it's one of the things that was working nicely with the minimum column width behaviour 🤔... I suppose it's hard to find the right balance! Here's a quick gif of dragging the column count:

![2023-04-28 10 45 08](https://user-images.githubusercontent.com/14988353/235028003-7e1749e0-90ff-4be2-98d7-cec5fa4b19fb.gif)

One other question from looking at this again: the `__unstableColumnCount` attribute is serialized to the block markup. If we keep the `__unstable` prefix, what would the process be to remove it further down the track if/when the attribute becomes stable? I don't think I've seen many other examples of migrated attributes within block supports, and if there's markup in the wild that has the `__unstable` prefix, will it mean we need the layout code to still work with the unstable version in perpetuity?

A couple of potential options:

* Add behaviour in the layout support so that attributes are grabbed with either the `__unstable` prefix or not, and use that as a kind of getter so that we can always preserve backwards compat for `__unstable` values in old post content
* Remove the `__unstable` prefix from being stored in block attributes, but still use the prefix as an API if we need it elsewhere?

Overall, though, this is feeling like a promising direction to me!

tellthemachines commented:
Thanks for the feedback @andrewserong !

> Is there a way we can introduce some kind of `minmax` rule at all to determine a cap on how many columns can be displayed at once depending on the size of the container?

I'm hesitant to introduce any smart behaviour here as it might mean people setting x amount of columns won't see that exact amount appear, and it can be confusing. But on trunk we currently have a limit of 6, and if we stick to that, it might not be too much of an issue?

> __unstableColumnCount attribute is serialized to the block markup. If we keep the __unstable prefix, what would the process be to remove it further down the track if/when the attribute becomes stable? I don't think I've seen many other examples of migrated attributes within block supports, and if there's markup in the wild that has the __unstable prefix, will it mean we need the layout code to still work with the unstable version in perpetuity?

Great point! I'm inclined to remove the `__unstable` altogether given that it's extremely unlikely we'll do away with the option of having a fixed number of columns - there are too many use cases where it would be handy.

andrewserong commented:
> I'm hesitant to introduce any smart behaviour here as it might mean people setting x amount of columns won't see that exact amount appear

Good point, and we can always look into smarter behaviour in follow-ups further down the track, too, rather than at this stage of working on fixing the Post Template block.

> But on trunk we currently have a limit of 6

Limiting to `6` columns in this context sounds like a good way to go 👍

tellthemachines committed:
Reduce max columns to 6.

tellthemachines committed:
Remove unstable prefix

tellthemachines commented:
Updated! If we're agreed on this approach being good, I'll start working on the deprecations.

andrewserong commented:
> Updated! If we're agreed on this approach being good, I'll start working on the deprecations.

This looks like a good direction to me, and it's testing pretty well so far! Do feel free to get more reviews / opinions if you're not sure, but so far this direction looks very promising to me.

At desktop viewport sizes column numbers 1 - 4 look nearly identical to trunk as far as I can tell. Once we go to 5 - 6 columns, it's a little curious that `grid` appears to cause the container to overflow, whereas the flex rules on trunk continue to force all the children to be within the space of the container (on TT3 at least), but that's not a blocker for pressing on with deprecations, particularly since the layout styles that get output can be changed independently of whatever's happening with the deprecations. Also, the example below is likely pretty artificial since it very clearly looks like a visual issue to a user in both cases:

| Five columns trunk | Five columns this PR |
| --- | --- |
| <img width="1359" alt="image" src="https://user-images.githubusercontent.com/14988353/235405712-bff76048-4132-4f54-8753-fda9836ebe37.png"> | <img width="1353" alt="image" src="https://user-images.githubusercontent.com/14988353/235405680-8e0e666e-d229-450e-bdbf-e4dac1b51f40.png"> |

tellthemachines commented:
> Do feel free to get more reviews / opinions if you're not sure, but so far this direction looks very promising to me.

I am very happy for more feedback if anyone wants to give it!

tellthemachines committed:
fix cols breaking out of container

tellthemachines commented:
> Once we go to 5 - 6 columns, it's a little curious that grid appears to cause the container to overflow

Turns out this one was an easy fix! It should look like trunk now. The front end is a bit broken for me because some of the legacy classnames are still being output; the deprecation should fix that.

ramonjd commented:
> Change the column width in the "grid" option and check that everything works as expected.

I'm a bit slow. While testing I couldn't for the life of me find a way to get the column width control to show. I see there's a [condition to show it where there are no columns](https://github.com/WordPress/gutenberg/pull/49050/files#diff-c4e4eb3411759ef40d434ab7b615cf55528dfbb1c8bdce4ec37f4927c2c7160aR48), but for all my post template blocks there's always a column count. 

The result is that the column width is always the default `12em`. 🤷 

Also, I tested mostly in empty theme, but in 2023 (with no global style modifications) I noticed that the layout (at least for me) didn't match the editor.

| 2023 editor  | 2023 frontend |
| ------------- | ------------- |
| <img width="400" alt="Screenshot 2023-05-02 at 12 43 28 pm" src="https://user-images.githubusercontent.com/6458278/235568463-2be1b9d1-d7a5-4d1b-bc15-ce1d47ffcaed.png">  | <img width="400" alt="Screenshot 2023-05-02 at 12 43 41 pm" src="https://user-images.githubusercontent.com/6458278/235568453-5d3e85c9-4baa-4c44-beed-5ed931f5d95b.png">  |

Is it just me?

> Adding responsive rules as part of layout will likely require modifying wp_style_engine_get_stylesheet_from_css_rules et al to deal with media queries in some way

Would CSS grid alone be an option for responsive layouts? I know that's a loaded question 😄 I was just playing with some approaches that can handle a max column count, but it assumes that we have fallback min column width and gap values. 

<details>

<summary>And to be honest, the CSS looks pretty gnarly</summary>

```php
$layout_styles[] = array(  
   'selector'     => $selector,  
   'declarations' => array(  
      'display'  => 'grid',  
      'grid-template-columns' => 'repeat(auto-fill, minmax(max(200px, calc((100% - calc(calc(' . $layout['columnCount'] . ' - 1) * var(--wp--preset--spacing--40))) / ' . $layout['columnCount'] . ')), 1fr))',  
   ),  
);
```

</details>


https://user-images.githubusercontent.com/6458278/235569498-074f8b29-7d56-4c58-b841-a0a2b20b5e11.mp4

tellthemachines commented:
Thanks for reviewing @ramonjd !

> I'm a bit slow. While testing I couldn't for the life of me find a way to get the column width control to show.

Sorry! This PR has been through multiple iterations and I didn't update the description after recent changes. You should see a number of columns field, not column width.

>  I noticed that the layout (at least for me) didn't match the editor.

Not it's not just you, this will be addressed by the deprecation. The problem is legacy classes are still being output on the server side.

> Would CSS grid alone be an option for responsive layouts?

The default grid that shipped in #49018 is responsive; it uses `auto-fill` and a user-changeable minimum column width. Using that here would break back compat though, because the current Post Template grid layout is based on user-set number of columns. That's what I'm trying to reproduce here.

ramonjd commented:
Ah great, thanks for the [explanations](https://github.com/WordPress/gutenberg/pull/49050#issuecomment-1530812823)! 🙇

tellthemachines committed:
Try adding deprecation

tellthemachines committed:
Fix broken loop

tellthemachines committed:
Update fixtures

tellthemachines commented:
Ok I've updated this PR to add the necessary deprecations, and everything seems to be working correctly. I've also updated the description to avoid further confusion 😄

ramonjd commented:
Following on from https://github.com/WordPress/gutenberg/pull/50079#issuecomment-1533905463, I can confirm that this branch works only by adding `wp_enqueue_global_styles` to `lib/compat/wordpress-6.3/get-global-styles-and-settings.php` (and removing/adding the appropriate filters).

Without it the grid CSS/classname isn't generated on the frontend. 

Looks like the [grid definitions haven't yet made it to core](https://github.com/ramonjd/wordpress-develop/blob/3895f70db0f4a51173ae6966d813fa5c6e8eb9ac/src/wp-includes/theme.json#L297).

So we still need to generate the stylesheet using `gutenberg_get_global_stylesheet()`, which uses `WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()`. 


<details>

<summary>lib/compat/wordpress-6.3/get-global-styles-and-settings.php changes</summary>

```diff
diff --git a/lib/compat/wordpress-6.3/get-global-styles-and-settings.php b/lib/compat/wordpress-6.3/get-global-styles-and-settings.php
index bb489664e1..e5a753cdf8 100644
--- a/lib/compat/wordpress-6.3/get-global-styles-and-settings.php
+++ b/lib/compat/wordpress-6.3/get-global-styles-and-settings.php
@@ -124,3 +124,54 @@ if ( ! function_exists( 'wp_get_block_css_selector' ) ) {
 function gutenberg_get_remote_theme_patterns() {
 	return WP_Theme_JSON_Resolver_Gutenberg::get_theme_data( array(), array( 'with_supports' => false ) )->get_patterns();
 }
+
+
+/**
+ * Enqueues the global styles defined via theme.json.
+ *
+ * @since 5.8.0
+ */
+function gutenberg_enqueue_global_styles() {
+	$separate_assets  = wp_should_load_separate_core_block_assets();
+	$is_block_theme   = wp_is_block_theme();
+	$is_classic_theme = ! $is_block_theme;
+
+	/*
+	 * Global styles should be printed in the head when loading all styles combined.
+	 * The footer should only be used to print global styles for classic themes with separate core assets enabled.
+	 *
+	 * See https://core.trac.wordpress.org/ticket/53494.
+	 */
+	if (
+		( $is_block_theme && doing_action( 'wp_footer' ) ) ||
+		( $is_classic_theme && doing_action( 'wp_footer' ) && ! $separate_assets ) ||
+		( $is_classic_theme && doing_action( 'wp_enqueue_scripts' ) && $separate_assets )
+	) {
+		return;
+	}
+
+	/*
+	 * If loading the CSS for each block separately, then load the theme.json CSS conditionally.
+	 * This removes the CSS from the global-styles stylesheet and adds it to the inline CSS for each block.
+	 * This filter must be registered before calling wp_get_global_stylesheet();
+	 */
+	add_filter( 'wp_theme_json_get_style_nodes', 'wp_filter_out_block_nodes' );
+
+	$stylesheet = gutenberg_get_global_stylesheet();
+
+	if ( empty( $stylesheet ) ) {
+		return;
+	}
+
+	wp_register_style( 'global-styles', false );
+	wp_add_inline_style( 'global-styles', $stylesheet );
+	wp_enqueue_style( 'global-styles' );
+
+	// Add each block as an inline css.
+	wp_add_global_styles_for_blocks();
+}
+
+remove_action( 'wp_enqueue_scripts', 'wp_enqueue_global_styles' );
+remove_action( 'wp_footer', 'wp_enqueue_global_styles', 1 );
+add_action( 'wp_enqueue_scripts', 'gutenberg_enqueue_global_styles' );
+add_action( 'wp_footer', 'gutenberg_enqueue_global_styles', 1 );


```

</details>

andrewserong commented:
> I can confirm that this branch works only by adding wp_enqueue_global_styles to lib/compat/wordpress-6.3/get-global-styles-and-settings.php (and removing/adding the appropriate filters).

Oh, thank you, I was just wondering why I couldn't get it working!

ramonjd commented on packages/block-editor/src/layouts/grid.js:
```diff
@@ -35,11 +35,23 @@ export default {
-				onChange={ onChange }
-			/>
+			<>
+				{ columnCount && (
```
Would this also work?

```jsx
		return layout?.columnCount ? (
			<GridLayoutColumnsControl layout={ layout } onChange={ onChange } />
		) : (
			<GridLayoutMinimumWidthControl
				layout={ layout }
				onChange={ onChange }
			/>
		);
```

tellthemachines replied:
Sure! A bit more succinct that way.

ramonjd commented on packages/block-editor/src/layouts/grid.js:
```diff
@@ -170,3 +186,23 @@ function GridLayoutMinimumWidthControl( { layout, onChange } ) {
+			label={ __( 'Columns' ) }
+			value={ columnCount }
+			onChange={ ( value ) => {
+				onChange( {
```
Optional without curlies:

```jsx
			onChange={ ( value ) =>
				onChange( {
					...layout,
					columnCount: value,
				} )
			}
```

ramonjd commented on packages/block-library/src/post-template/edit.js:
```diff
@@ -235,35 +237,68 @@ export default function PostTemplateEdit( {
-		</ul>
+		<>
+			<BlockControls>
+				{ ' ' }
```
Is this space intentional? If so, is it something CSS could take care of?

tellthemachines replied:
Oh. I'm 90% sure that was added by copilot and I somehow missed it 😳

ramonjd reviewed (commented):
This is working as advertised for me (assuming the relevant changes to global styles backend). Thanks for persevering!

I take it https://github.com/WordPress/gutenberg/pull/50310 or the equivalent is a dependency now...

I'm not a layout expert, so might wait for @andrewserong  to give it the ✅

andrewserong reviewed (commented):
This is testing pretty well for me so far, with the caveats of https://github.com/WordPress/gutenberg/pull/50310 not being rolled out yet. I like how the approach of this PR (and the deprecations) means that the flex-based unmigrated markup continues to function nicely with this PR applied, even before migrations have run — it feels like there's a good level of backwards compatibility ensured, there. Should we include a comment in the `style.scss` file for the block preceding the `&.is-flex-container` that those rules need to stay there for backwards compatibility?

Once https://github.com/WordPress/gutenberg/pull/50310 lands I'm happy to give this some more thorough testing!

tellthemachines commented:
Thanks for the feedback folks! Let's wait on #50310 for a final round of testing.

tellthemachines committed:
Code improvements

richtabor commented:
Is this ready for another review?

tellthemachines committed:
Add some comments to the CSS.

tellthemachines commented:
> Is this ready for another review?

It is now! #50310 is merged, I've rebased this branch and all the feedback so far has been addressed.

tellthemachines committed:
Add back deleted line

andrewserong reviewed (commented):
This is feeling very close to me! It's testing quite nicely on block themes for me:

✅ Existing query loop / post template blocks appear to render just as on trunk
✅ Updating an existing block by adjusting its number of columns transparently switches to the new grid approach, and it feels seamless from the user perspective to me
✅ Block spacing controls work nicely

https://github.com/WordPress/gutenberg/assets/14988353/114e8e7b-ff78-4268-9d42-034d6e1ddcb4

The only issue I ran into is that sometimes it appears that switching to Grid View, saving the editor, and then reloading, appears to forget the Grid View setting. Is it possible that one of the deprecations is accidentally firing when it shouldn't?

https://github.com/WordPress/gutenberg/assets/14988353/9305445b-a6ca-4074-be2c-b9e294e471c6

I noticed this while testing when I switched from my block theme and reloaded the editor in a classic theme, but I don't think it's related to classic themes as I could reproduce it with TT3. I'm not _too_ sure of the replication steps, but the blocks I'm working with in this example started out as (flex) grid blocks in `trunk`, then I made adjustments to them with this PR applied.

Other than that, this is looking really good, though! I'll just ping @ntsekouras for visibility, too, since I know Nik's been doing a bunch of work on the Query loop block lately, in case any of the changes here are of interest.

tellthemachines commented:
> I noticed this while testing when I switched from my block theme and reloaded the editor in a classic theme, but I don't think it's related to classic themes as I could reproduce it with TT3

How did you go about this? You shouldn't be able to access the site editor at all with a classic theme.

andrewserong commented:
> How did you go about this? You shouldn't be able to access the site editor at all with a classic theme.

It was a Query Loop block within a post in the post editor. A use-case example is classic themes that add Query Loop blocks that don't inherit the query from a template, to either posts or page (i.e. latest posts on an About or Articles page, for example). Classic themes still have access to all the Posts patterns, too, it looks like.

Another thing for us to test would be how it plays with plugins that extend the Query Loop block. I don't expect there'd be any issue with it, but the WooCommerce Products block appears to use Query Loop and Post Template blocks under the hood via a variation.

tellthemachines commented:
I can reproduce the issue with TT2 but _not_ with TT3. It might be related to the Query deprecation, but if so I'm not sure where it's getting the `displayLayout` value from - which is the condition for the migration to run - as it's removed during migration.

Update: yeah I think the bug only occurs when migrating from a Query that is set to display as list, which explains why I can repro it so easily on TT2. Still no idea how the displayLayout value is being preserved, unless it's via the block context.

andrewserong commented:
> Update: yeah I think the bug only occurs when migrating from a Query that is set to display as list, which explains why I can repro it so easily on TT2. Still no idea how the displayLayout value is being preserved, unless it's via the block context.

Yeah, it's very weird! This one has me pretty stumped for a Friday afternoon. I tried swapping out the `deprecated` value for an empty array `[]` and confirmed that that appears to resolve the issue, so it does sound like the problem is _somewhere_ in that file 😅 

https://github.com/WordPress/gutenberg/blob/c16dd860711e0a25da7966974bfb8cdfc7da7eea/packages/block-library/src/query/deprecated.js#L521

andrewserong commented:
Last hunch before I wrap up for the day:

I tried commenting out each of the deprecations one-by-one, and I think `v4` might potentially be the culprit? Its `isEligible` check will pass if there's no layout object set for the Query Loop. Then, I'm wondering if when it's running the `migrate` function, if it could be getting the default object value for the `displayLayout` attribute from its `attributes` key, set within the `v4` deprecation?

tellthemachines commented:
> Then, I'm wondering if when it's running the migrate function, if it could be getting the default object value for the displayLayout attribute from its attributes key, set within the v4 deprecation?

That could be the case, but still I'd expect legacy blocks to only go through the migration pipeline once. If they go again after being updated, then perhaps one of the isEligible checks isn't filtering out non-eligible blocks 🤔

tellthemachines commented:
Also I don't get why, not matter what I try, I can't reproduce this in TT3.

tellthemachines committed:
Remove legacy attributes in deprecation

tellthemachines commented:
Ok I think I fixed it. There were indeed some legacy layout properties slipping through in the v4 migration, which is why isEligible was being triggered repeatedly 😅 

Thanks for helping debug this @andrewserong !

ramonjd commented:
I've been testing this:

### Block themes (TT2 and TT3)
- Looks good in Firefox, Chrome and Safari
- Grid columns are consistent between front and backends
- I can confirm that the migration is taking place

![2023-05-29 12 03 09](https://github.com/WordPress/gutenberg/assets/6458278/66aa217f-1a07-4115-850f-14ead947d62c)



### Classic theme (TT1, and 2020)
- I can still reproduce the bug that @andrewserong [noted](https://github.com/WordPress/gutenberg/pull/49050#pullrequestreview-1444938665)
- To reproduce:
	1. On this branch, activate a classic theme (and have some published posts )
	2. Insert a query block in a post and switch to grid layout
	3. Publish/save
	4. Refresh the page
	5. Note the editor has reverted back to the list view
	6. The frontend, however, looks as it should (grid)

![2023-05-29 12 29 54](https://github.com/WordPress/gutenberg/assets/6458278/a76b3e3f-a4aa-4aa4-8489-692a3c2a1470)

I've been trying to track down the culprit. Before refreshing, the [layout hook](https://github.com/WordPress/gutenberg/blob/6297e8501ef2e5c3b0b50db19bb3b0cff890c821/packages/block-editor/src/hooks/layout.js#L347-L347) is detecting the correct layout and layout support.

Saving the post, I see that the correct attributes are being POSTED and saved in the database. 

The grid layout vanishes only after refreshing the page. The layout definitions are all available in [layout.js](https://github.com/WordPress/gutenberg/blob/6297e8501ef2e5c3b0b50db19bb3b0cff890c821/packages/block-editor/src/hooks/layout.js#L349-L349), but the incoming attributes revert from

```json
{
    "type": "grid",
    "columnCount": 3
}
```

to 

```json
{
    "type": "default"
}
```

They're not coming from the database, since the attributes are correctly saved:

<img width="731" alt="Screenshot 2023-05-29 at 12 59 25 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/8704e725-2f35-4e4a-9608-6bc8f8aa186a">

I wonder if there's a hook or something else running from Core that's parsing layout definitions from somewhere?

tellthemachines commented:
Hmmm, I'm not at all sure what's happening there except that it's not confined to classic themes: Query blocks added to both the post and the site editor with a block theme active are showing the same behaviour. 

I tried reverting the last commit and it didn't change anything (I didn't think it would), and this definitely wasn't happening on earlier stages of this branch so I suspect something came in with the latest rebase that is causing this issue.

One thing that I noticed  when testing with classic themes was that the spacing control isn't available, because the whole dimensions panel is hidden. It's not ideal, as spacing is the main practical reason for this PR 😅 but it's probably only fixable by unifying the layout controls as in #49459.

ramonjd commented:
>  it's not confined to classic themes: Query blocks added to both the post and the site editor with a block theme active are showing the same behaviour.

Oops! My bad for not trying both. It's a tricky one for sure. I eyeballed recent [commits](https://github.com/WordPress/gutenberg/commits/trunk?after=4bdc58146245014f58526922fb1b024a3c49ce50+34&branch=trunk&qualified_name=refs%2Fheads%2Ftrunk) and nothing stands out. A `git bisect` might be order?

tellthemachines commented:
bisect won't work unfortunately, because the branch was rebased 😞

tellthemachines committed:
Fix deprecation logic

tellthemachines commented:
Turns out it was another deprecation issue. Hopefully all is fixed now!

tellthemachines committed:
Update fixtures

andrewserong reviewed (commented):
This is feeling _so_ close, nice work tracking down those edge cases @tellthemachines! Just re-tested and it looks like the updates to the deprecations fixed the issues I was running into, and now updating the column count is preserved 👍

I also tested with WooCommerce, and it appears that this change plays nicely with the Products block and its variations for both Query Loop and Post Template. I.e. the deprecation is run, and it preserves the `__woocommerceNamespace` attributes as in the following migrated markup:

`<!-- wp:post-template {"layout":{"type":"grid","columnCount":4},"__woocommerceNamespace":"woocommerce/product-query/product-template"} -->`

Good stuff!

The last remaining issue I've run into is with default / fallback block spacing for the grid layout in classic themes. Currently in Classic themes, once the markup is updated and the default `1.25em` gap from the legacy `.is-flex-container` class is no longer applied, the grid layout renders no fallback gap value. Here's how it looks with a WooCommerce Products block:

<img width="677" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/a374991c-9f0f-4f14-adaa-ef3d9013f06c">

I _think_ the fix could be fairly straightforward. Over in the Post Template block's `block.json`, we can give the `blockGap` value a fallback value, similar to how the Columns block does it, by adding the following:

```
		"spacing": {
			"blockGap": {
				"__experimentalDefault": "1.5em"
			},
			"__experimentalDefaultControls": {
				"blockGap": true
			}
		}
```

Then, in the Theme JSON class, there's one line to update: https://github.com/WordPress/gutenberg/blob/03751d91dc7289d916b06f2010f28d25ee51167c/lib/class-wp-theme-json-gutenberg.php#L1301

With the grid layout now present, we could update this line to either `continue;` if we're looking at the default/flow layout, or add `&& 'grid' !== $layout_definition_key`.

With those two changes, in my local env it appears to restore the default fallback gap for the migrated markup:

<img width="673" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/a0ad4c0c-0251-485c-aff5-f89058adc031">

Would something like that work?

Edit: for consistency, I think we'd also need to update this line in the JS, too: https://github.com/WordPress/gutenberg/blob/81cd23380f5908022854bd0d367503bfb52ff9d7/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L451

tellthemachines committed:
Fallback gap for classic themes

tellthemachines committed:
fix spacing

tellthemachines commented:
Thanks for re-testing @andrewserong ! I was thinking we could revisit spacing support for classic themes once #49459 is underway again, but you're right that it's better to include a reasonable fallback for now. I copied the logic from the Columns block. 

Do we need that fallback in [use-global-styles-output.js](https://github.com/WordPress/gutenberg/blob/81cd23380f5908022854bd0d367503bfb52ff9d7/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L451)? That code never runs for classic themes; are there any other scenarios in which there's no block gap support? I'm assuming that in situations where it's explicitly disabled by a third party there will be some replacement value provided.

andrewserong commented:
Thanks for following up @tellthemachines! 

> Do we need that fallback in [use-global-styles-output.js](https://github.com/WordPress/gutenberg/blob/81cd23380f5908022854bd0d367503bfb52ff9d7/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L451)? That code never runs for classic themes; are there any other scenarios in which there's no block gap support?

The JS version gets runs for block themes that either don't opt-in to `appearanceTools: true`, or opt-in to `appearanceTools: true` but then also set `spacing.blockGap` to `null`, so I think we'll need that, too. I always forget because TT3 and a lot of blocks themes automatically opt in to appearanceTools / blockGap support 🙂

andrewserong commented on packages/block-library/src/post-template/block.json:
```diff
@@ -43,6 +41,14 @@
+		},
+		"spacing": {
+			"blockGap": {
+				"__experimentalDefault": "1.5rem"
```
Just double-checking, and it looks like the old `.is-flex-container` gap value is `1.25em`. Should we retain that for consistency, so that if someone goes to edit an existing page on a site without blockGap support, there's no change to the gap?

tellthemachines replied:
yeah that's a good point!

tellthemachines commented:
> The JS version gets runs for block themes that either don't opt-in to `appearanceTools: true`, or opt-in to `appearanceTools: true` but then also set `spacing.blockGap` to `null`

I was considering those as cases where it's explicitly disabled 😄 because folks who do that most likely want full control over the gap value, so having a fallback just means another default they have to override. But I added it in, anyway 🤷

tellthemachines committed:
match old default value

andrewserong commented:
> because folks who do that most likely want full control over the gap value, so having a fallback just means another default they have to override.

That's a good point. I'm mostly thinking of ensuring that the JS output in the site editor matches the site frontend, otherwise there'll be a discrepancy. But it is a good idea thinking through what the goal is for this type of user, as I'm sure it does cause a bit of friction — and we've largely been maintaining the `0.5em` gap for flex layout to ensure existing layouts (e.g. social icons) don't break anywhere... ah, the challenge of balancing back compat with how things should ideally work!! 😅

tellthemachines committed:
Update PHP test strings.

andrewserong commented on packages/block-library/src/query/edit/inspector-controls/index.js:
```diff
@@ -101,7 +101,7 @@ export default function QueryInspectorControls( props ) {
 	const showPostTypeControl =
 		! inherit && isControlAllowed( allowedControls, 'postType' );
-	const showColumnsControl = displayLayout?.type === 'flex';
+	const showColumnsControl = false;
```
This looks like a good and simple way to disable the control for now. I suppose in a follow-up we could remove the components from this file, but no need to do it as part of this PR.

andrewserong commented on test/integration/fixtures/blocks/core__query__deprecated-5.html:
```diff
@@ -0,0 +1,6 @@
+<!-- wp:query {"queryId":0,"query":{"perPage":10,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":true},"tagName":"main","displayLayout":{"type":"list"},"layout":{"type":"constrained"}} -->
+<div class="wp-block-query"><!-- wp:post-template -->
+    <!-- wp:post-title /-->
+    <!-- /wp:post-template -->
+</div>
+<!-- /wp:query -->
```
When copy/pasting this block markup into the editor it throws a block error because of the `div` tag (the block attributes specify a `tagName` of `main`).

The following markup migrates correctly for me:

```html
<!-- wp:query {"queryId":0,"query":{"perPage":10,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":true},"tagName":"main","displayLayout":{"type":"list"},"layout":{"type":"constrained"}} -->
<main class="wp-block-query"><!-- wp:post-template -->
    <!-- wp:post-title /-->
    <!-- /wp:post-template -->
</main>
<!-- /wp:query -->
```

tellthemachines replied:
Ooooh great attention to detail! I totally missed that 😂

andrewserong commented on test/integration/fixtures/blocks/core__query__deprecated-5.serialized.html:
```diff
@@ -0,0 +1,7 @@
+<!-- wp:post-title /-->
+<!-- /wp:post-template -->
+</div>
+<!-- /wp:query -->
```
This markup has not been migrated — the post template block doesn't receive the layout type. I believe this is because the source fixture has an error in it (the tag should be `main`, not `div`). When I update the markup locally, it appears to work well in the editor.

andrewserong reviewed (commented):
Great work @tellthemachines! Looks like there's an issue with one of the test fixtures that needs fixing, but other than that this looks like it should be good to go to me. 

✅ Unmigrated existing markup continues to render as on trunk
✅ Migrated markup continues to render correctly, matching the unmigrated state
✅ Controls at the Post Template level are working correctly, including block spacing
✅ 3rd party extension of Query Loop / Post Template block WooCommerce's Products block continues to work nicely with this update with one non-blocking exception (listed below)
✅ Classic themes output fallback styles correctly
❓ Block themes output fallback styles, however the order of the output means the `1.25em` doesn't take precedence

### Non-block issues (IMO)

In a blocks theme where `appearanceTools` is set to `false`, it appears that the order of styles being output results in the base blockGap value overriding the block-level one (so it gets `0.5em` instead of `1.25em`. This also affects the columns block, and it doesn't apply in classic themes, so I think this would be fine to look at separately:

<img width="1004" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/5612ff19-eb8a-40bd-a74d-5bbd80c0a2ac">

In the WooCommerce plugin, when inserting a Products block, they appear to use default attributes to set `displayLayout` to `flex` with `3` columns [here](https://github.com/woocommerce/woocommerce-blocks/blob/da191209e9ed5d6707132b168db7dacd427bdd7e/assets/js/blocks/product-query/constants.ts#L49. Possibly due to that attribute no longer existing, when a Products block is newly inserted, these attributes don't take effect so that default insertion experience is now a single column / default layout instead of 3 column layout:

![2023-06-02 10 37 18](https://github.com/WordPress/gutenberg/assets/14988353/1e77877f-30b4-4ead-b8fc-fdb00512cb59)

This only applies to newly inserted blocks, as existing blocks are correctly migrated. Because of this, I don't think this is a blocker as existing post content on current sites will be respected correctly, and users are still able to go into a newly inserted block and adjust the number of columns, etc. It does mean we should probably add the Needs Dev Note label, and in any dev note flag that plugins extending the Query Loop / Post Template blocks will need to update the attributes used if they're using custom default attributes on insertion.

---

All in all, other than the one deprecation fixture that needs updating, this looks in a good state to merge to me!

tellthemachines committed:
Fix tag discrepancy in fixtures

andrewserong reviewed (approved):
Thanks for fixing up the test fixtures, this LGTM now! ✨

tellthemachines commented:
> In a blocks theme where `appearanceTools` is set to `false`, it appears that the order of styles being output results in the base blockGap value overriding the block-level one (so it gets `0.5em` instead of `1.25em`. This also affects the columns block, and it doesn't apply in classic themes

I just opened #51187 to track this issue.

> This only applies to newly inserted blocks, as existing blocks are correctly migrated. Because of this, I don't think this is a blocker as existing post content on current sites will be respected correctly, and users are still able to go into a newly inserted block and adjust the number of columns, etc. It does mean we should probably add the Needs Dev Note label, and in any dev note flag that plugins extending the Query Loop / Post Template blocks will need to update the attributes used if they're using custom default attributes on insertion.

Good idea, I'll write up a dev note!

ntsekouras commented:
That's very nice to see! Thank you!

tellthemachines commented:
## Dev note

### Added layout and block spacing support to Post Template block

Previously, Post Template block had custom layout styles that allowed for either a "list" or a "grid" (implemented behind the scenes with CSS flex) layout, with controls living in its parent Query block.

For 6.3, layout and block spacing support have been added to Post Template, and its controls now live in the Post Template toolbar. There is still a choice of "list" and "grid" styles, but "grid" is now implemented with CSS grid.

fabiankaegy commented:
Stupid question, was this feature reverted from WordPress core at some point? 🤔 In 6.4 I only see the option to choose the number of columns. Not the minimum column width for some reason?

tellthemachines commented:
Not a stupid question at all @fabiankaegy! The grid option in Post Template has a slightly different configuration than the experimental Group grid variation. They both use grid layout, but this solution was thought best to preserve back compat, because Post Template always had the option to pick the number of columns. I think ideally there should be a way to switch between both grid configurations; with #42385 still in (slow) progress we haven't settled the best way to expose that in the UI.

fabiankaegy commented:
@tellthemachines Thanks for that clarification. Just to make sure, this means there also isn't a simple attribute one can change in a pattern for example to get the different grid behavior?

tellthemachines commented:
You can change the behaviour by editing the block attributes in the markup; the only thing that hasn't been implemented yet is the UI for switching between grid types. Essentially you'd need to replace the `columnCount` attrib with `minimumColumnWidth`.

So you'd end up with the Post Template layout attribute looking something like `"layout":{"type":"grid","minimumColumnWidth":"10rem"}`.

ObliviousHarmony commented:
@tellthemachines I'm working on a block with a requirement that responsiveness be implemented with a configurable number of maximum columns. This isn't something that layout supports, however, I _did_ come up with a solution by setting `minimumColumnWidth` to `max( 150px, (100% - ${blockGap * ( columns - 1 )}em) / ${ columns } )`. While this works, I am weary of shipping something that feels more like a consequence of missing validation than a deliberate decision.

Would you say that's a valid assumption, and if so, what would you suggest? I'm fine with opening up a PR if you agree supporting both `columnCount` and `minimumColumnWidth` makes sense. We have an L-2 WordPress support policy though so it would be great to have an interim solution.

tellthemachines commented:
@ObliviousHarmony `columnCount` and `minimumColumnWidth` should be mutually exclusive: the grid is either configured with a fixed number of columns, or with a fixed minimum width and adjustable number of columns. What you're describing sounds like it would work better as a fixed number of columns with perhaps some container queries to adjust that number as needed on smaller screens/containers. 

Container query based responsiveness is [under discussion](https://github.com/WordPress/gutenberg/issues/19909#issuecomment-2024471271) as a feature to add to core, but it will take a while to ship so my suggestion in the meantime would be to add some custom styles to your block in a way that can be removed later once core support is in place.

ObliviousHarmony commented:
Thanks for getting back to me so quickly @tellthemachines.

Sorry, I wrote this comment _way_ too early in the morning and shared very little information; I even gave some CSS without any context 😅. I appreciate your suggestion here, however, a container query doesn't quite meet my requirements. The feedback we've received has steered us towards having a maximum number of columns that decreases the number of columns if one of them would fall below a minimum width. We're currently migrating a Query Loop like block from `displayLayout` to `layout` and previously used column class styles like this:

```css
grid-template-columns: repeat( auto-fill, minmax( max( 150px, ( 100% - 3.75em ) / 4 ), 1fr ) );
```

The approach basically just boils down to using a percentage width (with a minimum) to select the maximum number of columns. When it can't fit that many columns it creates new rows and decreases the number of columns accordingly. This is much simpler than a container query. It provides an easy user experience in the form of a single slider rather than potentially complicated configuration using container sizes.

I don't think `columnCount` and `minimumColumnWidth` are necessarily mutually exclusive because they're describing two different concepts. One is a constraint that you'd like a specific number of columns while the other is a constraint that you'd like for your columns to have a minimum width. It feels reasonable to assume you could combine these two constraints to create a more constrained but still responsive grid.

As noted, I _do_ have a working implementation by passing part of the above style to `minimumColumnWidth`. This just feels like a hack since it doesn't seem intentional that you can pass CSS functions to something described as a "width" :smile:

tellthemachines commented:
@ObliviousHarmony that's an interesting use case!

> It provides an easy user experience in the form of a single slider rather than potentially complicated configuration using container sizes.

So which variable would you allow users to configure, column count or minimum width? I wasn't thinking of exposing the container query configuration to users; rather it could be a dynamic query based on the column count/min width values such as [we already have](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L648) for grid span resizing.

ObliviousHarmony commented:
> So which variable would you allow users to configure, column count or minimum width?

In our case we would be letting them set the column count and handle setting the minimum width internally.

> I wasn't thinking of exposing the container query configuration to users; rather it could be a dynamic query based on the column count/min width values such as [we already have](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L648) for grid span resizing.

Even so, it still feels like one of the powerful aspects of a CSS Grid is the ability for the browser to handle column overflow behavior. Since that's the specific kind of responsiveness I'm suggesting here it seems like a reasonable feature to add. What do you think? Would you be opposed to me opening a PR adding a case for both `columnCount` and `minimumColumnWidth` being set?

tellthemachines commented:
> Since the `layout` control is delegated to the implementing block it seems reasonable that developers could decide how to utilize it?

The primary goal for layout (and other block supports) is to provide a way to configure layout options in the UI. Block developers can define a layout type and choose to expose the controls or not, and theme authors can further choose to hide layout controls altogether or for specific blocks.

The way block supports are built relies on a close correspondence between settings such as `columnCount` and `minimumColumnWidth` and specific UI controls. In the case of these two settings, the presence of one or the other determines which controls are exposed for a given block, and it's possible to toggle between them with the Manual/Auto buttons in the layout panel. For that reason, it's not viable to use `minimumColumnWidth` as a way for developers to configure custom responsive behaviour when `columnCount` is also present.

It's very likely that at some point additional responsiveness will be added to the "Manual" grid option, but it's not yet clear how that might be exposed in the UI or as an API for developers. That should be a part of the wider work on responsive controls being discussed in #19909. The reason I mentioned container queries above is partly because the discussion over there is leaning towards them, but also because responsive styles for grid need to take into account the behaviour of multi-span children, and once #59483 is moved out of experimental status, blocks with fixed positions in the grid. Container queries are a good way of ensuring consistent behaviour between the parent grid and its children on resize.

ObliviousHarmony commented:
In our design the control is expressed as the number of maximum columns. With the context you've provided, how would you feel about a new `maximumColumnCount` option to pair with `minimumColumnWidth`? When set it would apply a `grid-template-columns` style like the one I shared earlier. This would be a new control for the "Auto" mode.

tellthemachines commented:
> With the context you've provided, how would you feel about a new `maximumColumnCount` option to pair with `minimumColumnWidth`?

@ObliviousHarmony this is a timely comment, because #60941 was started yesterday proposing something similar 😄  

It's a bit buggy right now because the logic to toggle between Auto and Manual modes is still looking at `columnCount` but if you switch it to Auto you should be able to test the behaviour.
 
For context, that draft resulted from [this conversation](https://github.com/WordPress/gutenberg/pull/59490#issuecomment-2068383744), where we're looking at changing Manual mode to more easily allow for positioning blocks in fixed points in the grid. Feedback is welcome in either place!

ObliviousHarmony commented:
> this is a timely comment, because https://github.com/WordPress/gutenberg/pull/60941 was started yesterday proposing something similar

This _is_ timely. I wrote essentially the exact same thing, even down to using the presence of `minimumColumnWidth` in the layout to indicate whether or not the count should be static or responsive. Thanks @tellthemachines, I'll leave feedback there. Really excited about the Grid block too, just not something I can use for this purpose!
